File: agent/common/crypto.py
# termhop agent — session-key handshake and AEAD primitives. No hand-rolled
# crypto: X25519 ECDH and XChaCha20-Poly1305 come from pynacl (libsodium
# bindings), HKDF comes from `cryptography` (pynacl has no first-class HKDF
# and hand-rolling one from raw HMAC would be error-prone). Each library is
# used for the primitive it implements directly.
#
# Pinned protocol-v2 parameters:
#   - ECDH: X25519 via nacl.bindings.crypto_scalarmult.
#   - Authentication: a 256-bit out-of-band pairing secret is the HKDF salt;
#     the canonical transcript hash is included in HKDF info.
#   - KDF output: independent agent->client, client->agent, and proof keys.
#   - AEAD: XChaCha20-Poly1305 (nacl.bindings, IETF construction), 24-byte
#     nonce plus AAD binding direction, type, session_id, and sequence.
import base64
import hashlib
import hmac
import os
from dataclasses import dataclass

import nacl.bindings
import nacl.utils
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

_HKDF_INFO_PREFIX = b"termhop-handshake-v2\x00"
_SESSION_KEY_LEN = 32  # bytes — required by crypto_aead_xchacha20poly1305_ietf
_KEY_MATERIAL_LEN = _SESSION_KEY_LEN * 3
_NONCE_LEN = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES


class CryptoError(Exception):
    pass


@dataclass(frozen=True)
class SessionKeys:
    agent_to_client: bytes
    client_to_agent: bytes
    proof: bytes


def generate_ephemeral_keypair() -> tuple[bytes, bytes]:
    """Returns (private_key_raw, public_key_raw), each 32 bytes."""
    private_key = nacl.utils.random(nacl.bindings.crypto_scalarmult_SCALARBYTES)
    public_key = nacl.bindings.crypto_scalarmult_base(private_key)
    return private_key, public_key


def handshake_transcript(
    *, session_id: str, token: str, agent_pubkey_b64: str, client_pubkey_b64: str
) -> bytes:
    """Canonical transcript shared byte-for-byte with client/src/lib/crypto.js."""
    return (
        "termhop-handshake-v2\n"
        f"session_id={session_id}\n"
        f"token={token}\n"
        f"agent_pubkey={agent_pubkey_b64}\n"
        f"client_pubkey={client_pubkey_b64}"
    ).encode("ascii")


def _decode_pairing_secret(pairing_secret: str) -> bytes:
    try:
        padding = "=" * (-len(pairing_secret) % 4)
        raw = base64.b64decode(pairing_secret + padding, altchars=b"-_", validate=True)
    except (ValueError, TypeError) as exc:
        raise CryptoError(f"invalid pairing secret: {exc}") from exc
    if len(raw) != 32:
        raise CryptoError("pairing secret must decode to exactly 32 bytes")
    return raw


def derive_session_keys(
    own_private_key: bytes,
    peer_public_key: bytes,
    pairing_secret: str,
    transcript: bytes,
) -> SessionKeys:
    """Authenticated ECDH + HKDF -> directional traffic and proof keys."""
    if len(own_private_key) != 32 or len(peer_public_key) != 32:
        raise CryptoError("X25519 private and public keys must each be 32 bytes")
    try:
        shared_secret = nacl.bindings.crypto_scalarmult(own_private_key, peer_public_key)
    except Exception as exc:
        raise CryptoError(f"ECDH failed: {exc}") from exc

    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=_KEY_MATERIAL_LEN,
        salt=_decode_pairing_secret(pairing_secret),
        info=_HKDF_INFO_PREFIX + hashlib.sha256(transcript).digest(),
    )
    material = hkdf.derive(shared_secret)
    return SessionKeys(
        agent_to_client=material[:_SESSION_KEY_LEN],
        client_to_agent=material[_SESSION_KEY_LEN : _SESSION_KEY_LEN * 2],
        proof=material[_SESSION_KEY_LEN * 2 :],
    )


def pairing_proof(proof_key: bytes, transcript: bytes, role: str) -> str:
    if role not in {"agent", "client"}:
        raise CryptoError(f"invalid proof role: {role}")
    proof = hmac.new(proof_key, transcript + b"\nrole=" + role.encode("ascii"), hashlib.sha256).digest()
    return base64.b64encode(proof).decode("ascii")


def verify_pairing_proof(proof_key: bytes, transcript: bytes, role: str, proof_b64: str) -> bool:
    return hmac.compare_digest(pairing_proof(proof_key, transcript, role), proof_b64)


def message_aad(*, type_: str, session_id: str, seq: int, direction: str) -> bytes:
    if direction not in {"agent_to_client", "client_to_agent"}:
        raise CryptoError(f"invalid message direction: {direction}")
    return (
        "termhop-message-v2\n"
        f"direction={direction}\n"
        f"type={type_}\n"
        f"session_id={session_id}\n"
        f"seq={seq}"
    ).encode("ascii")


def encrypt(key: bytes, plaintext: bytes, *, aad: bytes) -> tuple[str, str]:
    """Returns (nonce_b64, ciphertext_b64). Ciphertext includes the
    Poly1305 tag (libsodium's combined-mode API) — no separate tag field."""
    nonce = os.urandom(_NONCE_LEN)
    ciphertext = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_encrypt(
        plaintext, aad, nonce, key
    )
    return base64.b64encode(nonce).decode("ascii"), base64.b64encode(ciphertext).decode("ascii")


def decrypt(key: bytes, nonce_b64: str, ciphertext_b64: str, *, aad: bytes) -> bytes:
    try:
        nonce = base64.b64decode(nonce_b64, validate=True)
        ciphertext = base64.b64decode(ciphertext_b64, validate=True)
    except (ValueError, TypeError) as exc:
        raise CryptoError(f"invalid base64 in encrypted payload: {exc}") from exc
    if len(nonce) != _NONCE_LEN:
        raise CryptoError(f"nonce must decode to exactly {_NONCE_LEN} bytes")
    if len(ciphertext) < nacl.bindings.crypto_aead_xchacha20poly1305_ietf_ABYTES:
        raise CryptoError("ciphertext is shorter than the authentication tag")

    try:
        return nacl.bindings.crypto_aead_xchacha20poly1305_ietf_decrypt(
            ciphertext, aad, nonce, key
        )
    except Exception as exc:
        raise CryptoError(f"decryption/authentication failed: {exc}") from exc


def encode_pubkey(pubkey_raw: bytes) -> str:
    return base64.b64encode(pubkey_raw).decode("ascii")


def decode_pubkey(pubkey_b64: str) -> bytes:
    try:
        raw = base64.b64decode(pubkey_b64, validate=True)
    except (ValueError, TypeError) as exc:
        raise CryptoError(f"invalid base64 pubkey: {exc}") from exc
    if len(raw) != 32:
        raise CryptoError("public key must decode to exactly 32 bytes")
    return raw

File: client/src/lib/crypto.js
// termhop client — session-key handshake and AEAD primitives, mirroring
// agent/common/crypto.py 1:1. No hand-rolled crypto: X25519 ECDH and
// XChaCha20-Poly1305 from @noble/curves/@noble/ciphers (audited, pure JS —
// same precisely-specified standards as the Python side's pynacl/libsodium
// binding, not a "different library" risk since these are RFC 7748/RFC 8439
// primitives). HKDF from the browser's native SubtleCrypto — RFC 5869
// like Python's `cryptography` HKDF.
//
// Pinned protocol-v2 parameters (must match agent/common/crypto.py exactly):
//   - KDF: HKDF-SHA256 with the 256-bit out-of-band pairing secret as salt
//     and the canonical transcript hash in info.
//   - Output: independent agent->client, client->agent, and proof keys.
//   - AEAD: XChaCha20-Poly1305, 24-byte random nonce per message, combined
//     ciphertext+tag, with direction/type/session/sequence bound as AAD.
import { xchacha20poly1305 } from "@noble/ciphers/chacha";
import { x25519 } from "@noble/curves/ed25519";

const HKDF_INFO_PREFIX = new TextEncoder().encode("termhop-handshake-v2\0");
const KEY_MATERIAL_LEN_BITS = 768;
const NONCE_LEN = 24;

export class CryptoError extends Error {}

export function generateEphemeralKeypair() {
  const privateKey = x25519.utils.randomSecretKey();
  const publicKey = x25519.getPublicKey(privateKey);
  return { privateKey, publicKey };
}

export function handshakeTranscript({ sessionId, token, agentPubkeyB64, clientPubkeyB64 }) {
  return new TextEncoder().encode(
    `termhop-handshake-v2\nsession_id=${sessionId}\ntoken=${token}\nagent_pubkey=${agentPubkeyB64}\nclient_pubkey=${clientPubkeyB64}`
  );
}

export async function deriveSessionKeys(ownPrivateKey, peerPublicKey, pairingSecret, transcript) {
  let sharedSecret;
  try {
    sharedSecret = x25519.getSharedSecret(ownPrivateKey, peerPublicKey);
  } catch (err) {
    throw new CryptoError(`ECDH failed: ${err.message}`);
  }

  const transcriptHash = new Uint8Array(await crypto.subtle.digest("SHA-256", transcript));
  const info = new Uint8Array(HKDF_INFO_PREFIX.length + transcriptHash.length);
  info.set(HKDF_INFO_PREFIX);
  info.set(transcriptHash, HKDF_INFO_PREFIX.length);
  const salt = fromBase64Url(pairingSecret);
  if (salt.length !== 32) throw new CryptoError("pairing secret must decode to exactly 32 bytes");

  const keyMaterial = await crypto.subtle.importKey("raw", sharedSecret, "HKDF", false, ["deriveBits"]);
  const bits = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt, info },
    keyMaterial,
    KEY_MATERIAL_LEN_BITS
  );
  const material = new Uint8Array(bits);
  return {
    agentToClient: material.slice(0, 32),
    clientToAgent: material.slice(32, 64),
    proof: material.slice(64, 96),
  };
}

export async function pairingProof(proofKey, transcript, role) {
  if (role !== "agent" && role !== "client") throw new CryptoError(`invalid proof role: ${role}`);
  const suffix = new TextEncoder().encode(`\nrole=${role}`);
  const message = new Uint8Array(transcript.length + suffix.length);
  message.set(transcript);
  message.set(suffix, transcript.length);
  const hmacKey = await crypto.subtle.importKey("raw", proofKey, { name: "HMAC", hash: "SHA-256" }, false, ["sign"]);
  return toBase64(new Uint8Array(await crypto.subtle.sign("HMAC", hmacKey, message)));
}

export async function verifyPairingProof(proofKey, transcript, role, proofB64) {
  const expected = fromBase64(await pairingProof(proofKey, transcript, role));
  const actual = fromBase64(proofB64);
  if (expected.length !== actual.length) return false;
  let different = 0;
  for (let i = 0; i < expected.length; i++) different |= expected[i] ^ actual[i];
  return different === 0;
}

export function messageAad({ type, sessionId, seq, direction }) {
  if (direction !== "agent_to_client" && direction !== "client_to_agent") {
    throw new CryptoError(`invalid message direction: ${direction}`);
  }
  return new TextEncoder().encode(
    `termhop-message-v2\ndirection=${direction}\ntype=${type}\nsession_id=${sessionId}\nseq=${seq}`
  );
}

export function encrypt(key, plaintext, aad) {
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LEN));
  const cipher = xchacha20poly1305(key, nonce, aad);
  const ciphertext = cipher.encrypt(plaintext);
  return { nonceB64: toBase64(nonce), ciphertextB64: toBase64(ciphertext) };
}

export function decrypt(key, nonceB64, ciphertextB64, aad) {
  let nonce, ciphertext;
  try {
    nonce = fromBase64(nonceB64);
    ciphertext = fromBase64(ciphertextB64);
  } catch (err) {
    throw new CryptoError(`invalid base64 in encrypted payload: ${err.message}`);
  }

  try {
    const cipher = xchacha20poly1305(key, nonce, aad);
    return cipher.decrypt(ciphertext);
  } catch (err) {
    throw new CryptoError(`decryption/authentication failed: ${err.message}`);
  }
}

export function encodePubkey(pubkeyRaw) {
  return toBase64(pubkeyRaw);
}

export function decodePubkey(pubkeyB64) {
  try {
    const key = fromBase64(pubkeyB64);
    if (key.length !== 32) throw new CryptoError("public key must decode to exactly 32 bytes");
    return key;
  } catch (err) {
    throw new CryptoError(`invalid base64 pubkey: ${err.message}`);
  }
}

export async function sessionFingerprint(sessionKeys, transcript) {
  if (!sessionKeys || !transcript) return null;
  const label = new TextEncoder().encode("termhop-session-fingerprint-v2\0");
  const material = new Uint8Array(label.length + transcript.length + 64);
  material.set(label);
  material.set(transcript, label.length);
  material.set(sessionKeys.agentToClient, label.length + transcript.length);
  material.set(sessionKeys.clientToAgent, label.length + transcript.length + 32);
  const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", material));
  return Array.from(digest.slice(0, 6), (byte) => byte.toString(16).padStart(2, "0").toUpperCase())
    .join("")
    .match(/.{1,4}/g)
    .join(" ");
}

function toBase64(bytes) {
  return btoa(String.fromCharCode(...bytes));
}

function fromBase64(b64) {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

function fromBase64Url(value) {
  const normalized = value.replace(/-/g, "+").replace(/_/g, "/");
  return fromBase64(normalized + "=".repeat((4 - (normalized.length % 4)) % 4));
}
